### ISSUE TITLE
Constant time base64

### ISSUE THREAD

**Issue by DemiMarie** [briansmith/ring#559]
This is needed to avoid a timing leak when reading and writing PEM encoded structures.

**Comment by briansmith** [comment]
I agree, but the problems can be avoided much more simply by not using base64 at all. *ring* intentionally doesn't support any base64 functionality as of now. OTOH it wouldn't be hard to add constant-time base64 to *ring* using its internal constant-time primitives, but it might be problematic to export enough of the private constant-time primitives so that others can implement base64 themselves on top of *ring*, so I'd at least review a PR that adds this.

**Comment by DemiMarie** [comment]
I agree that using base64 for secrets is a bad idea.  On the other hand, a
lot of formats require it.

My current thought is to start with a simple implementation, then perhaps
later add optimized implementations that use SIMD instructions.  I suspect
that SIMD should be a big win.

**Comment by briansmith** [comment]
I don't think there's any need for a SIMD implementation for secrets, as secrets are small.
